urlcaef331f4e_dreamable_cra_tgz-pJ8-coho_lp_project_LPProjectJ8/src/coho/interp/MatrixValue.java

```diff
@@ -0,0 +1,970 @@
+					+ ", c0 + src.ncols = " + (c0 + src.size(1)));
+		Matrix x = m.convert();// BasicMatrix.convert(m); XXX clone
+		x.assign(m, 0, 0);
+		x.assign(src, r0, c0);
+		return (new MatrixValue(x));
+	}
+
+	public static MatrixValue matrix(RCvalue args, String who,
+			CohoType elementType) throws EvalException {
+		if (args.size() != 1)
+			throw new EvalException("usage:  " + who + "(row or column)");
+		if (!(args.value(0) instanceof RCvalue))
+			throw new EvalException(who + ":  argument must be a row or column");
+		return (new MatrixValue((RCvalue) (args.value(0)), elementType));
+	}
+
+	public static Enumeration matrixFns() {
+		final Function[] functions = new Function[] { new Function() {
+			public String name() {
+				return ("matrix");
+			}
+
+			public Value eval(RCvalue args) throws EvalException {
+				return (matrix(args, name(), CohoDouble.type));// CohoDouble?
+			}
+		}, new Function() {
+			public String name() {
+				return ("boolMatrix");
+			}
+
+			public Value eval(RCvalue args) throws EvalException {
+				return (matrix(args, name(), CohoBoolean.type));
+			}
+		}, new Function() {
+			public String name() {
+				return ("intMatrix");
+			}
+
+			public Value eval(RCvalue args) throws EvalException {
+				return (matrix(args, name(), CohoInteger.type));
+			}
+		}, new Function() {
+			public String name() {
+				return ("size");
+			}
+
+			public Value eval(RCvalue args) throws EvalException {
+				int[] sz = matrix_arg(args, 0, name()).size();
+				Matrix m = IntegerMatrix.create(2, 1);// BasicMatrix.create(2,
+				// 1, CohoInteger.type);
+				m.assign(sz[0], 0);
+				m.assign(sz[1], 1);
+				return (new MatrixValue(m));
+			}
+		}, new Function() {
+			public String name() {//XXX
+				return ("range");
+			}
+
+			public Value eval(RCvalue args) throws EvalException {
+				if (args.size() != 2)
+					throw new EvalException("usage:  " + name() + "(lo, hi)");
+				if (!((args.value(0) instanceof DoubleValue) && (args.value(1) instanceof DoubleValue)))
+					throw new EvalException(name()+ ":  both arguments must be doubles");
+				return (new RangeValue(
+						(int) (Math.round(((DoubleValue) (args.value(0))).value())), 
+						(int) (Math.round(((DoubleValue) (args.value(1))).value()))
+				));
+			}
+		}, new Function() {
+			public String name() {
+				return ("transpose");
+			}
+
+			public Value eval(RCvalue args) throws EvalException {
+				if (args.size() != 1)
+					throw new EvalException("usage:  " + name() + "(matrix)");
+				return (new MatrixValue(matrix_arg(args, 0, name()).transpose()));
+			}
+		}, new Function() {
+			public String name() {
+				return ("diag");
+			}
+
+			public Value eval(RCvalue args) throws EvalException {
+				if (args.size() != 1)
+					throw new EvalException("usage:  " + name() + "(vector)");
+				Matrix m = matrix_arg(args, 0, name());
+				if (m.isVector()) {
+					try {
+						return (new MatrixValue(m.diag()));
+					} catch (MatrixError e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)